### DISCUSSION TITLE
Handling many navigation tabs?

### DISCUSSION THREAD

**Discussion by bigpick** [squidfunk/mkdocs-material#4765] [Q&A] [answered]
# Info

When using the [navigation tabs][] feature, top level sections get rendered in the menu layer below the header as expected.

However, if you get too many top level sections, the  list of tabs that is rendered into the menu layer ends up getting cut off at some point, and there seems to be no way to scroll/move/access any rest of the remaining tabs.

Has anyone messed around with any custom tweaks to make the rendering of many top level tabs better when using the feature, along with many top level sections? Or is the guidance simply "if you have that many top level sections, don't use navigation tabs"? I'd hope not the latter, since I really enjoy having the tabs on larger viewports, and would be great to be able to leverage it still, but understand if so. 

I've searched around this repo's discussions/issues/PRs, and the only thing I could find related was a discussion on getting the tab section items themselves to have dropwdowns (which was [struck  down]) which wouldn't help in this case.

## Reproduction

```bash
mkdir /tmp/testsite
cd /tmp/testsite

cat << EOF > mkdocs.yml
site_name: test many navs
theme:
  name: material
  features:
    - navigation.tabs

nav:
  - Home:
    - index.md
    - Developing this site locally: devel.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
  - About this site:
    - theme/index.md
    - Markdown spec: theme/kitchen_sink.md
    - Additional rich syntax: theme/mkdocs_additions.md
  - Foobar:
    - theme/index.md
EOF

mkdir -p docs/theme

for dummy in index.md theme/index.md theme/kitchen_sink.md theme/mkdocs_additions.md devel.md; do touch docs/${dummy}; done

mkdocs serve
```

![Screenshot 2022-12-15 at 9 30 05 AM](https://user-images.githubusercontent.com/9803299/207887469-daab77a0-9776-451e-87f8-4442d1028069.png)

Thanks!

[navigation tabs]: <https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#navigation-tabs>
[struck down]: <https://github.com/squidfunk/mkdocs-material/discussions/2198#discussioncomment-438284>

**Comment by bigpick** [comment]
Reproduction site example as a zip instead of manual commands: 

* [testsite.zip](https://github.com/squidfunk/mkdocs-material/files/10241625/testsite.zip)

**Reply by squidfunk** [reply]
Thanks for providing the details and the example, very valueable. How do you imagine so many tab working, i.e. what would be from your perspective be the ideal behavior? Overflowing? Scrollable? Should they extend to the right?

**Reply by bigpick** [reply]
Hm, I suppose horizontal scrolling would be most ideal, all things considered. Messing about with it, horizontal scrolling seems to be the best option to me. Though, not sure how to handle the scrollbar styling, as the default option looks quite ugly and in the way, imo:

e.g.: by simply adding 
```css
.md-tabs__list {
...
  white-space: nowrap;
  overflow-x: auto;
...
}
```

Can see what I mean: 
![Screenshot 2022-12-16 at 1 53 05 PM](https://user-images.githubusercontent.com/9803299/208168844-39cac362-0de4-4c41-996a-b20ad217fbce.png)

In my limited experience, I vaguely remember scrollbar styling not being very reliable and/or cross-browser compatible, though, so not sure if it would be feasible.

**Comment by squidfunk** [comment, marked as answer]
While this is definitely an edge case, I took the opportunity and fixed the issue at hand in 9fb36102d, allowing the navigation tabs container to scroll. Adding so many top-level sections that the container overflows is discouraged, since you're hiding content on a horizontal axis that might be harder to reach on some devices. For that reason, the fix isn't beautiful, but it should do the trick.

The fix will be released with v9, tracked in #4714. You can try it now by installing the latest beta, see issue.

**Comment by CroMarmot** [comment]
> While this is definitely an edge case, I took the opportunity and fixed the issue at hand in [9fb3610](https://github.com/squidfunk/mkdocs-material/commit/9fb36102dfadd4e02e21bbba47d3667f11b67479), allowing the navigation tabs container to scroll. Adding so many top-level sections that the container overflows is discouraged, since you're hiding content on a horizontal axis that might be harder to reach on some devices. For that reason, the fix isn't beautiful, but it should do the trick.
> 
> The fix will be released with v9, tracked in #4714. You can try it now by installing the latest beta, see issue.

Is this supported?

<img width="1794" height="578" alt="Image" src="https://github.com/user-attachments/assets/7b35262c-e4c9-455e-97ed-2144c39f19b6" />

<img width="1723" height="544" alt="Image" src="https://github.com/user-attachments/assets/05fed482-a6c8-4179-9b50-531df38b2b12" />

envrionment:

windows chrome138.0.7204.169 (64bit)
windows firefox developer edition 142.0b4 (64-bit)
windows edge 131.0.2903.86

reproduce: open https://www.paddleocr.ai/latest/version3.x/pipeline_usage/OCR.html in desktop browser


their github ci https://github.com/PaddlePaddle/PaddleOCR/actions/runs/15952598110/job/44994609699

<img width="971" height="701" alt="image" src="https://github.com/user-attachments/assets/c81c34dc-b400-499d-b5a9-b29c885626f4" />

**Reply by squidfunk** [reply]
Yes, it works, the tabs are scrollable, exactly as noted in my comment:

> While this is definitely an edge case, I took the opportunity and fixed the issue at hand in [9fb3610](https://github.com/squidfunk/mkdocs-material/commit/9fb36102dfadd4e02e21bbba47d3667f11b67479), allowing the navigation tabs container to scroll. Adding so many top-level sections that the container overflows is discouraged, since you're hiding content on a horizontal axis that might be harder to reach on some devices. For that reason, the fix isn't beautiful, but it should do the trick

**Reply by kamilkrzyskow** [reply]
Scrollbar is not shown, but you can scroll it.
Press Shift + use scroll wheel on the mouse.
or press middle mouse button between the tabs to enable scrolling mode and scroll horizontally by moving the mouse.

This is mostly for mobile usage cases with touch screen. The scrollbar is hidden on purpose:
https://github.com/squidfunk/mkdocs-material/blob/9d958543d01ccedd0b6531f8129cfb76ef3d812a/src/templates/assets/stylesheets/main/components/_tabs.scss#L64-L68

You could restore it using customization:
https://squidfunk.github.io/mkdocs-material/customization/#additional-css

**Reply by squidfunk** [reply]
Ah, the scrollbar, thanks @kamilkrzyskow. Mobile devices don't show the navigation tabs anyway, but yes, we still consider this an edge case that can be overwritten with simple customization, if needed. The situation could definitely be improved with scrollable arrows like we do for content tabs, but since it's an edge case, it's not something we can currently invest time into.

**Reply by CroMarmot** [reply]
`Press Shift + use scroll wheel on the mouse` and `press middle mouse button` works for me! thanks.

**Reply by DingJunyao** [reply]
Multi-rows nagavation tabs may be better. Fewer users know how to scroll horizontally than those who know how to scroll vertically.

And I hope the author can add this feature.

**Reply by kamilkrzyskow** [reply]
Hi @DingJunyao
you can add it with CSS customization, maybe limit it with a media query to PC width, as Tablet users might be used to scrolling via touch, but idk, as Tablet mode activates also for some 1400x900 small Mac screens even though it's mouse and keyboard.

https://squidfunk.github.io/mkdocs-material/customization/#additional-css

```css
.md-tabs__list {
  flex-wrap: wrap;
}
```
<details open>
<summary> Example </summary>

<img width="1678" height="973" alt="image" src="https://github.com/user-attachments/assets/4a7673b3-18a2-4223-b358-d9d2dc1f0698" />

</details>
